build-tools/automation/build.groovy

```diff
@@ -121,12 +178,34 @@ timestamps {
 
         stageWithTimeout('run all tests', 160, 'MINUTES', XADir, false) {   // Typically takes 1hr and 50 minutes (or 110 minutes)
             echo "running tests"
-            def stageStatus = sh(
-                script: "make run-all-tests CONFIGURATION=${env.BuildFlavor}",
-                returnStatus: true
-            );
 
-            if (stageStatus != 0) {
+            def commandStatus = 0
+
+            if (isPr) {
+                def hasPrLabelFullMonoIntegrationBuild = hasPrLabel(env.GitRepo, env.ghprbPullId, 'full-mono-integration-build')
+                def hasPrLabelRunTestsRelease = hasPrLabel(env.GitRepo, env.ghprbPullId, 'run-tests-release')
+
+                commandStatus = sh(
+                    script: """
+                        # If PR has the 'full-mono-integration-build' or 'run-tests-release' label, run w/ SKIP_NUNIT_TESTS set
+                        if ${hasPrLabelFullMonoIntegrationBuild} || ${hasPrLabelRunTestsRelease}; then
+                            echo "Run all tests: The 'full-mono-integration-build' and/or 'run-tests-release' labels have been found on the PR"
+                            make run-all-tests CONFIGURATION=${env.BuildFlavor} SKIP_NUNIT_TESTS=1
+                        else
+                            echo "Run all tests: Neither of the 'full-mono-integration-build' or 'run-tests-release' labels have been found on the PR"
+                            make run-all-tests CONFIGURATION=${env.BuildFlavor}
+                        fi
+                    """,
+                    returnStatus: true
+                )
+            } else {
+                commandStatus = sh(
+                    script: "make run-all-tests CONFIGURATION=${env.BuildFlavor}",
+                    returnStatus: true
+                )
+            }
```

Comment:
I'd refactor this whole thing as:

```groovy
def skipNunitTests = false

if (isPr) {
  def hasPrLabelFullMonoIntegrationBuild = hasPrLabel(env.GitRepo, env.ghprbPullId, 'full-mono-integration-build')
  def hasPrLabelRunTestsRelease = hasPrLabel(env.GitRepo, env.ghprbPullId, 'run-tests-release')
  skipNunitTests = hasPrLabelFullMonoIntegrationBuild || hasPrLabelRunTestsRelease
  echo "Run all tests: Labels on the PR: 'full-mono-integration-build' (${hasPrLabelFullMonoIntegrationBuild}) and/or 'run-tests-release' (${hasPrLabelRunTestsRelease})"
}

if (skipNunitTests) {
  commandStatus = sh (script: "run-all-tests CONFIGURATION=${env.BuildFlavor} SKIP_NUNIT_TESTS=1", returnStatus: true)
} else {
  commandStatus = sh (script: sh "make run-all-tests CONFIGURATION=${env.BuildFlavor}", returnStatus: true)
}

Reply:
Refactored per your suggestion. commit https://github.com/xamarin/xamarin-android/pull/2767/commits/3a001aa1ef8417097260e275b5a46bd93be85b9f